Use scikit-learn confusion matrix instead of own custom code

jo-mueller opened the pull request:
# Enhanced overlap matrix calculation with sklearn implementation

The calculation of the label overlap matrix has been exchanged with the implementation from [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html), which is considerably faster. After the upgrade, the benchmark (/notebooks/label/overlap_matrix_benchmarking.ipynb)) shows comparable results for both approaches.

## Type of change

* [ ] Bug-fix
* [x] New feature
* [ ] Breaking change
* [ ] Documentation update

## References 

Closes #3 

## Tests

* [ ] I adapted existing tests, because
* [ ] I added new tests to cover the code change
* [x] All tests pass with my change

## Final checks

* [ ] My change is the minimal possible work for the desired feature/fix
* [x] I updated the documentation where necessary to cover the change

haesleinhuepf commented on biapol_utilities/label/_label_overlap_matrix.py:
```diff
@@ -17,21 +19,17 @@ def label_overlap_matrix(label_image_x, label_image_y):
-    # preallocate a 'contact map' matrix
-    overlap = np.zeros((1 + label_image_x.max(), 1 + label_image_y.max()), dtype=np.uint)
+    # Calculate confusion matrix
+    overlap = metrics.confusion_matrix(label_image_x, label_image_y)
```
Is the function `label_overlap_matrix(a, b)` really necessary, if we also could just call `sklearn.metrics.confusion_matrix(a.ravel(), b.ravel())`? See also my comment in https://github.com/BiAPoL/biapol-utilities/issues/3#issuecomment-962593032.

I explicitly suggest to remove `label_overlap_matrix`. Let me know what speaks against that.

jo-mueller replied:
Hm...
+ Makes code cleaner and shorter
- If we eventually decide to replace this function by something faster/more appropriate for superlarge images, i.e. GPU-acceleration, we would have to re-introduce it.

haesleinhuepf replied:
Then let's remove it for now and potentially later introduce it in case we want to GPU-accelerate something.

haesleinhuepf commented on biapol_utilities/label/_label_overlap_matrix.py:
```diff
@@ -17,21 +19,17 @@ def label_overlap_matrix(label_image_x, label_image_y):
-#     y = (utils.format_labels(y)).ravel()
+    
+    
+    # Flatten input arrays
```
```suggestion
    # Make input arrays 1D
```
... in ImageJ "Flatten" means something very different. That's why I'm suggesting to be more precise here.

haesleinhuepf commented on biapol_utilities/label/_match_labels.py:
```diff
@@ -61,6 +62,10 @@ def match_labels(label_image_x, label_image_y, method=intersection_over_union, *
         threshold = kwargs.get('iou_threshold', 0.25)
     
+    # relabel label_image_y to keep overlap matrix small
+    label_image_y, _, _ = relabel_sequential(label_image_y)
```
Great solution. I fully agree that here in this context relabeling makes sense. :-)

jo-mueller replied:
Thanks! :)

haesleinhuepf commented on biapol_utilities/label/_match_labels.py:
```diff
@@ -61,6 +62,10 @@ def match_labels(label_image_x, label_image_y, method=intersection_over_union, *
     
+    # relabel label_image_y to keep overlap matrix small
+    label_image_y, _, _ = relabel_sequential(label_image_y)
+    # label_image_x, fw, inv = relabel_sequential(label_image_x)
```
```suggestion
```

haesleinhuepf commented on tests/test_intersection_over_union.py:
```diff
@@ -13,18 +13,17 @@ def test_iou():
     iou = biau.label.intersection_over_union(a, b)
     
-    assert(iou.shape == (size, size))
+    assert(iou.shape == (len(a), len(b)))
```
```suggestion
    assert(iou.shape == (np.max(a), np.max(b)))
```
Just a quick question: Would this also work and be more general? I mean with any label images and not just with `range()`-made label images?

jo-mueller replied:
I *think* the generalized solution would be something like
```python
assert(iou.shape == (len(np.unique(a)), len(np.unique(b)))
```

I'll try around a bit.

jo-mueller replied:
I think this would be a more generalized test:

```python
    # Create two arrays of random length
    size = np.random.randint(5, 10)
    low = 1
    a = np.arange(low, size, 1, dtype=np.uint8)
    b = np.arange(low, size, 1, dtype=np.uint8)
    
    # Shuffle array
    np.random.shuffle(b)
    
    # delete random entries from a and b
    idx = np.random.randint(low, size-1, 1)
    b[idx] = 0
    a[idx] = 0
    
    iou = biau.label.intersection_over_union_matrix(a, b)
    
    assert(iou.shape == (len(np.unique([a, b])), len(np.unique([a, b]))))
```

As I see it, `confusion_matrix` from sklearn will always return a square matrix, even if the number of unique labels is different in the two input labels. Depending on the present labels, there will be some empty columns, which should be handled by `relabel_sequential` in https://github.com/BiAPoL/biapol-utilities/pull/20#discussion_r745445866.

As in https://github.com/BiAPoL/biapol-utilities/pull/20#discussion_r745442795, this is a bit of a tautological test right now, but could become useful if we eventually decide to re-implement the confusion_matrix calculation.

haesleinhuepf reviewed (changes requested):
Hi Johannes @jo-mueller ,

just some minor stuff and I'm suggesting to remove one function. Furthermore, could you please also delete `overlap_matrix_benchmarking.ipynb`? I created this branch with that notebook just to demonstrate something. I wasn't thinking of making this notebook part of the repository.

I'm happy to merge your changes afterwards.

Thanks!

Best,
Robert

jo-mueller commented:
Synopsis:

- Removed label_overlap_matrix (https://github.com/BiAPoL/biapol-utilities/pull/20/commits/ad0e7f806c962c18d0843c92bf08f329c20c4052)
- generalized IoU test (https://github.com/BiAPoL/biapol-utilities/pull/20/commits/cd894ae383c338343063fbc9355d704cb4f3b536)
- removed benchmark notebook (56e718f45c00a77ef7bf5ce3e61b4ecf947a830c)

haesleinhuepf commented on tests/test_intersection_over_union.py:
```diff
@@ -4,23 +4,36 @@
     assert np.allclose(iou, reference, 0.02)
+    
+if __name__ == "__main__":
+    test_iou()
```
Please add a line break by the end of every file.
![image](https://user-images.githubusercontent.com/12660498/141107474-25eeb221-71e3-4888-a0d1-18bc9f8041c6.png)

jo-mueller replied:
Changed in 3c94fb2090ab463cd761e67c331a8298a0169127

haesleinhuepf commented:
Hi Johannes @jo-mueller ,

this all looks good to me. Shall I merge?

jo-mueller commented:
Yes :)

haesleinhuepf commented:
Wonderful. Thanks for working on this @jo-mueller !
